### ISSUE TITLE
API interactivity

### ISSUE THREAD

**Issue by LuiggiTenorioK** [BSC-ES/autosubmit-api#55]
Summarizing the discussion about the goal of this year, we are aiming to add more interactive endpoints to the API that will allow users to trigger actions that can modify the state of the experiments. To accomplish this there are some issues we have to solve:

### Define the scope

We need to list requirements to better structure the changes we want to make. In this particular case, we can list the actions (run experiment, update description, change status, etc) we desire to include inside the API. Then, we can make a formal endpoint definition in OpenAPI with the route, expected request, and response.

Also, this will help us to link the effort we have to do in other tasks (DDBB sync, security, communication with Autosubmit, ...)

**[UPDATE]** Mapped actions until now:

* Start Experiment `POST /v4/experiments/<expid>?action=run` -> `start`
* Stop Experiment `POST /v4/experiments/<expid>?action=stop` -> `stop`
* Set Status to Job `PATCH /v4/experiments/<expid>/jobs/<jobid>?status=<newstatus>` -> `setstatus`
* Create experiment `POST /v4/experiments` -> `expid`
* Generate the experiment `POST /v4/experiments/<expid>?action=generate` -> `create`
* Restart the experiment `POST /v4/experiments/<expid>?action=restart` -> `recovery`

### Set some infrastructure cases

There are multiple scenarios in which the API will be installed like in ES, [Climate DT](https://earth.bsc.es/gitlab/digital-twins/de_340/project_management/-/issues/538#note_248597), and EDITO. Is important that we formally define those to better understand the bounds (security, network, dependencies) we are going to have for each one.

### Define the action procedures

As discussed, there are some options to process the actions we want to include inside the API:

* **Modify the artifacts (DDBB and files) directly:** This option is great for actions that update the metadata but will require a better understanding of the data sources as discussed in #53 to unify definitions and responsibilities. Might be hard to ensure retro compatibility as different versions of Autosubmit might use different data sources.
* **Call the Autosubmit commands:** This can be used for actions that need to execute extensive procedures through a subprocess. Also, it will not add a dependency directly to the API. This allows the possibility to run the commands even if the API and Autosubmit are in different nodes but we will need to figure out how to ensure a safe communication between them. There is some discussion about it in autosubmitreact#21.
* **Use Autosubmit as a Python package:** In this case, Autosubmit can be called directly through the code and can be more easily handled. Still, it adds a hard requirement that Autosubmit should be a dependency of the API. In both, the previous and this option, retrocompatibility is delegated to Autosubmit.

@mcastril @kinow

**Comment by LuiggiTenorioK** [comment]
Right! There are different problems to solve. Adding the interactive endpoints is a must for sure. On the other hand, we have to find a way to handle experiments of different versions in different environments.

For the different versions issue, I think Miguel is right that we can use the `-v` flag to solve it.

For the different environment issues, there are the daemon synchronization and the higher-level API solutions. IMO the synchronization solution is way more complex, especially considering the different versions of the experiments. Then, having a higher-level API will work better for EDITO as it will only need one service from where the requests will be made (SURF). 

This higher-level API solution is inspired by another project used in the most popular workflow manager in bioinformatics (https://galaxyproject.org/) which uses a similar lower-level API called [Pulsar](https://github.com/galaxyproject/pulsar) to solve the same problem we have.

**Comment by LuiggiTenorioK** [comment]
Our problem is more or less stated here: https://pulsar.readthedocs.io/en/latest/containers.html

(I remember stating similar issues in my Master's thesis)

**Comment by LuiggiTenorioK** [comment]
From today's meeting:

![20240201_150049](https://earth.bsc.es/gitlab/es/autosubmit-api/uploads/9c539cc7a4f48c97b83d21e89dbfba86/20240201_150049.jpg)

**Comment by LuiggiTenorioK** [comment]
In GitLab by @kinow on Feb 2, 2024, 10:38

Thank you for attaching it here, @LuiggiTenorioK !

@mcastril , should one of us get in touch with EDITO/SURF to schedule a meeting to discuss about this? If so, maybe it would be Quentin/Renaud and Francesco from CMCC via email explaining what we want to discuss and asking for best time/day for the meeting? Thank you

**Comment by LuiggiTenorioK** [comment]
In GitLab by @mcastril on Feb 7, 2024, 12:33

Yes Bruno, thanks for volunteering. Please refer to Renaud, Quentin and Francesco together, with us in copy.

**Comment by LuiggiTenorioK** [comment]
In GitLab by @kinow on Feb 13, 2024, 11:57

Meeting doodle poll sent.

Summarizing what we discussed in the meeting.

- Are we going to have a single API instance, shared by all the EDITO-Infra users, or are we going to have one API per user? Or both?
- If we have instances/containers for the API shared by users, which user could we use to run the API? Would it have access to S3 and to an SSH key to connect to HPC or other EDITO-Infra instances?
- How are the API and GUI instances going to be started? By a user action, or Kubernetes will keep a minimum number of pod(s) running?
- We need to define how/if we will use a shared file-system for Autosubmit experiments. For the demo we used S3, but that's not really an option for Autosubmit (we use NFS at ClimateDT and BSC). Maybe we could allocate a permanent volume to be bound to each Autosubmit container (with enough storage for the experiments? GB's, TB's?)
- Do we have a list of requirements, or how SURF GUI will interact with the API? This would be useful to validate the endpoints we will have to implement.

**Comment by LuiggiTenorioK** [comment]
In GitLab by @kinow on Feb 29, 2024, 11:26

>Do we have a list of requirements, or how SURF GUI will interact with the API? This would be useful to validate the endpoints we will have to implement.

Use the API to request the status of the experiment that is running. We can build a list of experiments the user is running, and show how long the experiment is taking, resources used...

When the user defines what they want to submit, at that point they submit the list of tasks and the run of the job starts.

Users must be able to restart from a certain point in the workflow (`setstatus`).

Q: Can we get the list of `N` last experiments?

Yes, but the deployment option could define how it works.

Possible endpoints required:

- list experiments per user
  - get experiment details (if not returned in the previous call)
- create experiments (from scratch or from a template)
- launch an experiment
- restart an experiment
- stop experiment

**Comment by LuiggiTenorioK** [comment]
In GitLab by @kinow on Feb 29, 2024, 11:52

>Are we going to have a single API instance, shared by all the EDITO-Infra users, or are we going to have one API per user? Or both?

Both ways are possible, but it's on the business side to choose.

- Surf GUI
- Surf API (?)
- Autosubmit API - python process
- Autosubmit GUI - HTML httpd
- Autosubmit - python process, reading a shared DB

We need to choose between Process and Service (in EDITO Infra).

In a Service you can launch multiple tools (GUI + AS API + Surf API + etc).

Not tested, but it should be possible to have dependencies between Services.

In Datalab we have "Projects". We could create the project "Edito ModelLab". You can create an instance with members of the project. You can also share the URL of the project with members outside the project.

> If we have instances/containers for the API shared by users, which user could we use to run the API? Would it have access to S3 and to an SSH key to connect to HPC or other EDITO-Infra instances?

This needs to be tested to confirm. It should be possible to share the instance so others can manage it too.

> How are the API and GUI instances going to be started? By a user action, or Kubernetes will keep a minimum number of pod(s) running?

In the project we can have services that are always available. At the moment +2 weeks old services are killed, but this may change in the future.

Q: Who maintains the infra (if a service goes down?)

Suggestion: use replication (pods/etc) to have more resources, configure the helm/etc to have higher availability.

Q: Surf GUI can use the API

**Comment by LuiggiTenorioK** [comment]
In GitLab by @kinow on Feb 29, 2024, 12:00

Q: Can I deploy to another service/catalogue/env?

At the moment merge requests go to production. Staging is for EDITO-Infra. EDITO-Infra team is working to give others access to the playground catalogue/env.

N.B.: BSC team to use the playground. Then later we ask it to be moved to the modellab/ai/etc catalogue.

>We need to define how/if we will use a shared file-system for Autosubmit experiments. For the demo we used S3, but that's not really an option for Autosubmit (we use NFS at ClimateDT and BSC). Maybe we could allocate a permanent volume to be bound to each Autosubmit container (with enough storage for the experiments? GB's, TB's?)

At the moment this is not doable. But that should be possible under the common modellab project. So all users under that project can target that volume there. We can test that after the modellab is created.

Q: are we going to give access to external users, to use this database (on the shared docker volume)?

...

**Comment by LuiggiTenorioK** [comment]
In GitLab by @kinow on Feb 29, 2024, 12:04

Action: we need to define who are the users of the model lab too. At the moment anyone can request an account to EDITO. The Catalogue Service view is available to unauthenticated users.

**Comment by LuiggiTenorioK** [comment]
changed due date to December 31, 2024

**Comment by LuiggiTenorioK** [comment]
Coming back to the API interactivity, I was thinking about the **start experiment** feature. I remember some options that we discussed earlier last year, so I did a pro/cons list for each of them to better decide which is the best approach. Here they are:

---

## Option 1: API has Autosubmit as a Python dependency and uses the Python interface

The API will install a version of Autosubmit in their pip dependencies, then it will use the `Autosubmit.run_experiment()` method to run the experiment.

Pros:

 - Everything is in Python
 - Easier to implement

Cons:

 - All experiments will be updated to the version the API has. No retrocompatibility.
 - All the experiment workflows will be executed in a single instance
 - API needs access to the persistence layer (FS/DB)
 - Experiments can only be stopped by the API
 
 
## Option 2: API makes ssh to another node 

API will use paramiko to execute `autosubmit run -v`. The user will need to provide the SSH info in the GUI to run the experiment. This info can include commands to be executed before (e.g. `module load autosubmit:4.1.11`).


Pros:

 - Workload can be distributed
 - Users will use their OS user
 - Users can manually stop their workflow by using their SSH info provided

Cons:

 - A lot of info needs to be provided (give private key, host, module load) by the user side
 - Security should be highly inspected (prevent injections or any other vulnerability)
 
 
## Option 3: API can register runners and use them

This is option 2 but persistent the SSH info (host, private key) of the dedicated nodes set by the maintainer of the API. Users will select in which node (pre-defined) they want to run their workflow.

Pros:

 - Setup only once per runner
 - User-friendly
 - Workload could be distributed

Cons:

 - Depends on a sysadmin to configure the runners
 - Security has to be highly checked to not expose sensitive information
 - Experiments can only be stopped by the API (unless the user has access to the runners and permission to kill that process)
 
---

@kinow @dbeltrankyl @mcastril WDYT? Do you have any other option in mind or pro/con that is not considered?

**Comment by LuiggiTenorioK** [comment]
From today's meeting:

Option 2 can be improved in this way:

* Users will receive the public key of the API node, so they will need to give the API permission to access their node. Then, users will give their node's hostname to the API.
* Detailed setup instructions should be given in the GUI and documentation (RtD)
* API should know which API users can trigger that experiment which has to be set by the experiment user (Experiment whitelist)

**Comment by LuiggiTenorioK** [comment]
About how to implement the run experiment, I'm first introducing 2 concepts in the API:

### Runner

This is the mechanism by which the API will run the "write" operations, handling the execution in a selected node. The following runners will be implemented:

- `LocalRunner`: This runner handles the execution on the same node by creating subprocesses
- `SSHRunner`: Handles the execution on an SSH passwordless host (TBD)
- `ContainerRunner`: Execution in existing Docker/Podman containers (TBD)

### ModuleLoader

To make the preparations for further Autosubmit commands, there are some common ways to do them. These loaders will load the Autosubmit environment in a structured way, preventing command injection risk. These loaders are planned to be implemented:

- `CondaModuleLoader`: Load conda for executing commands via `conda run`.
- `LmodModuleLoader`: Load using `module load`. It can handle multiple modules.
- `VenvModuleLoader`: Load from a path in the host the Python virtual envs using `source <path>`.
- `NoModuleLoader`: Direct execution without loading anything.

----

During the first implementation, `LocalRunner` will be available with all the `ModuleLoaders`. Then, inside the API, we will probably handle which runners/loaders can be used with some config file for the API (probably a topic of another issue).

In the case of running an experiment in the API, the runner that is currently used should be persisted to handle a stop experiment command.

Note: For the future, we should also implement some kind of permission system per resource to manage a set of users that can execute commands per experiment.

**Comment by LuiggiTenorioK** [comment]
API runners (#192) have been merged into the main branch of the API as an alpha feature.

As next steps to improve this feature, we have to:

- [ ] Implement the `SSHRunner`
- [ ] Validate support of `LmodModuleLoader`
- [ ] Test compatibility with systems without `/bin/bash`
- [ ] Add `expid` and `create` commands in runners
- [ ] Use `autosubmit stop` to stop the experiments

Maybe we can discuss the priority of these improvements in the next meeting.

**Comment by ntorqulu** [comment]
p.d. @LuiggiTenorioK I'm not sure on the priority of this one.

Several sub-issues will be opened to address this one:

1. Implement restart/recovery -> I think it's the only command not implemented listed in this issue.
2. Per-experiment authorization -> currently any authenticated user can trigger runners on any expid. Needs an auth check (e.g. needs a dependency that verifies that the caller is the exp owner or in an allowed list (defined by the owner?)) before `run/stop/setstatus/create`.
3. Command-injections -> make it safer, stop handling a string to a shell (shell=True) for a list of arguments with no shell. Avoid possible injections.
4. ssh-keys policy -> replace `AutoAddPolicy()` with one that can be verified and configured by the user.
5. Move `v4alpha` endpoints to `v4`

Probably there's more, this would be sub-issues of this parent one...

**Comment by LuiggiTenorioK** [comment]
According to the current project's needs (at least until the end of this year), those issues are not high priority, even if some of them sound like it.

> * Implement restart/recovery -> I think it's the only command not implemented listed in this issue.

Restart doesn't exist, and recovery hasn't been requested yet, as far as I know.

> * Per-experiment authorization -> currently any authenticated user can trigger runners on any expid. Needs an auth check (e.g. needs a dependency that verifies that the caller is the exp owner or in an allowed list (defined by the owner?)) before `run/stop/setstatus/create`.

Related to #70. However, it feels more like an Autosubmit feature than an API one.

> * Command-injections -> make it safer, stop handling a string to a shell (shell=True) for a list of arguments with no shell. Avoid possible injections.

I think they have to use shell to use bash and load the modules properly. So, not sure we can stop handling outside a shell.

> * ssh-keys policy -> replace `AutoAddPolicy()` with one that can be verified and configured by the user.

Feasibility hasn't been evaluated.

> * Move `v4alpha` endpoints to `v4`

All of them have been moved to `v4` with changes to how they worked. I think they can be removed now that the EDITO project has finished, but I say it may be better to do it when the new service is updated for the MOIRAI project.

**Comment by ntorqulu** [comment]
Got it, thanks!
